learning_observer/learning_observer/google.py

```diff
@@ -404,73 +94,168 @@ def get_error_details(error):
+        )
+        return courses
+
+    @register_cleaner("document", "doctext", GOOGLE_ENDPOINTS)
+    def extract_text_from_google_doc_json(
+            j, align=True,
+            EXTRACT_DEBUG_CHECKS=False):
+        '''
+        Extract text from a Google Docs JSON object, ignoring formatting.
 
+        There is an alignment issue between Google's and Python's handling
+        of Unicode. We can either:
+        * extract text faithfully (align=False)
+        * extract text with aligned indexes by cutting text / adding
+        spaces (align=True)
 
-@register_cleaner("coursework_submissions", "assigned_docs")
-def clean_assignment_docs(google_json):
-    '''
-    Retrieve set of documents per student associated with an assignment
-    '''
-    student_submissions = google_json.get('studentSubmissions', [])
-    for student_json in student_submissions:
-        google_id = student_json[constants.USER_ID]
-        local_id = learning_observer.auth.google_id_to_user_id(google_id)
-        student_json[constants.USER_ID] = local_id
-        docs = [d['driveFile'] for d in learning_observer.util.get_nested_dict_value(student_json, 'assignmentSubmission.attachments', []) if 'driveFile' in d]
-        student_json['documents'] = docs
-        # TODO we should probably remove some of the keys provided
-    return student_submissions
+        This issue came up in text with a Russian flag unicode symbol
+        (referencing the current conflict). I tried various encodings,
+        and none quite matched Google 100%.
 
+        Note that align=True doesn't necessarily give perfect local alignment
+        within text chunks, since we do have different lengths for something like
+        this flag. It does work okay globally.
+        '''
+        # return error message for text
+        if 'error' in j:
+            return get_error_details(j['error'])
+        length = j['body']['content'][-1]['endIndex']
+        elements = [a.get('paragraph', {}).get('elements', []) for a in j['body']['content']]
+        flat = sum(elements, [])
+        text_chunks = [f.get('textRun', {}).get('content', '') for f in flat]
+        if align:
+            lengths = [f['endIndex'] - f['startIndex'] for f in flat]
+            text_chunks = [_force_text_length(chunk, length) for chunk, length in zip(text_chunks, lengths)]
+        text = ''.join(text_chunks)
+
+        if EXTRACT_DEBUG_CHECKS:
+            print("Text length versus Google length:")
+            print(len(text), length)
+            print("We expect these to be off by one, since Google seems to starts at 1 (and Python at 0)")
+            if align:
+                print
+                print("Offsets (these should match):")
+                print(list(zip(itertools.accumulate(map(len, text_chunks)), itertools.accumulate(lengths))))
+
+        return {'text': text}
+
+    @register_cleaner("coursework_submissions", "assigned_docs", GOOGLE_ENDPOINTS)
```

Comment:
We are passing the same parameter to each of the register cleaner functions. Create a closure or use the `functools.partial` function to set it once.

Reply:
I've added this